src/compiler/checker.ts

```diff
@@ -4420,12 +4420,26 @@ namespace ts {
             let container = getThisContainer(node, /*includeArrowFunctions*/ false);
             let parent = container && container.parent;
             if (parent && (isClassLike(parent) || parent.kind === SyntaxKind.InterfaceDeclaration)) {
-                if (!(container.flags & NodeFlags.Static)) {
+                if (!(container.flags & NodeFlags.Static) && !isConstructorParameter(node, container)) {
                     return getDeclaredTypeOfClassOrInterface(getSymbolOfNode(parent)).thisType;
                 }
             }
             error(node, Diagnostics.A_this_type_is_available_only_in_a_non_static_member_of_a_class_or_interface);
             return unknownType;
+
+            function isConstructorParameter(node: Node, container: Node) {

```

Comment:
Just as an aside, nested local functions like this are more expensive than sibling functions because a new function object has to be created every time the outer function is invoked. But in this particular case, I'd just get rid of this function altogether.